src/processors/default/processMempool.ts

```diff
@@ -0,0 +1,64 @@
+
+
+import _ from 'lodash';
+
+import { Table } from '../../db/db';
+import { pendingMempoolMessages } from '../../triggers/messages';
+import { CrdtOpetation, CrdtState, MempoolMessage, PendingMempoolMessage } from '../../types/message';
+import { Clients, Processor } from '../../types/processor';
+
+export const processMempool: (table: Table) => Processor = 
+  (table) => ({
+
+    name: 'processMempool',
+    trigger: pendingMempoolMessages(table),
+    processorFunction: async (messages: PendingMempoolMessage[] , clients: Clients) => {
+      // ignore inserts for now
+      const staleMessages: MempoolMessage[] = messages.filter(message => message.operation === CrdtOpetation.INSERT);
+
+      // messages are ordered by timestamp.
+      //  group by table, column, entity so that we can easily categorize fresh and stale messages
+      const groupedMessagesUpdates = _.groupBy(
+        messages.filter(message => message.operation === CrdtOpetation.UPDATE),
+        message => `${message.operation}.${message.table}.${message.column}.${message.entityId}`
+      );
+      const entityUpdates: { [id: string]: { [column: string]: string } } = {};
+      const crdtUpdates: CrdtState[] = [];
+
+      Object.values(groupedMessagesUpdates).forEach(groupMessage => {
+        // gets the freshest message out of the mempool
+        const freshestUpdate = groupMessage.pop()!;
```

Comment:
It would be nice if we could remove the assumption that timestamps don't clash and that clients have synchronised clocks. This would also simplify your zora-approvals migration so that you can remove all the delay stuff. One easy way we could do this right here is to implement a really simple hybrid logical clock that just does the following:
 - filter updates to find all the updates that have the freshest timestamp
 - if theres only 1, great, use it and continue
 - if there are more than 1, sort them by a string sort on the value and use the first one

Reply:
This is built with the assumption that all clients have synchronised clocks. The delays stuff in the zora-approvals migration is just so I can test messages with new and old timestamps, but that won't be necessary in production.

I understand that using a string sort on the value will get us a way to resolve conflicts into a single update, but how can we be sure that that result is the one we want (the one that happened last)? Wouldn't we need something separate to the value to determine the latest message? E.g a lamport clock

Reply:
Well what I'm suggesting is that we're making the choice to say that actually, the first message in the string sort IS the last message and IS the one we want.

So in the case of a clashing timestamp, we're saying we always do want the first value in the string sort.

The only risk/edge I think is if there is some potential edge case scenario where we incidentally/accidentally create 2 updates on the same column with the same timestamp and actually do expect that there is a specific order for them to be processed in? Is there any scenario in which you imagine this happening?